rustls/src/lib.rs

```diff
@@ -637,6 +637,41 @@ pub mod manual;
 
 pub mod time_provider;
 
+/// APIs for implementing locks on no-std environments.
+pub mod lock {
+    #[cfg(not(feature = "std"))]
+    use alloc::boxed::Box;
+
+    #[cfg(not(feature = "std"))]
+    /// A lock protecting shared data.
+    pub trait Lock: Send + Sync {
```

Comment:
Like `TimeProvider` I feel like this is another candidate for living in another crate, like `abstractions-that-should-be-in-core`. In no sane world is a TLS library where abstractions over OS fundamentals should live, especially if they're gratuitously under-specified like these.

My other thought on this is -- can we instead use `spin` and avoid this entirely? That would be a similar solution to using hashbrown.

Reply:
>  can we instead use spin and avoid this entirely? That would be a similar solution to using hashbrown.

That sounds appealing. I can look at that option.

Reply:
> In no sane world is a TLS library where abstractions over OS fundamentals should live, especially if they're gratuitously under-specified like these.

Just found https://docs.rs/lock_api/0.4.11/lock_api/

Reply:
:bulb: